### PR TITLE
Tagged Metrics for TimeLock: Clients and Leadership

### PR DESCRIPTION
**Pull request by jeremyk-91**
- [x] Needs smoketest/verification.

**Goals (and why)**:
- Current timelock server-side metrics are an aggregate across all clients, and don't expose leadership information. This makes reading and debugging timelock metrics cleaner and easier. See internal ticket 138.

**Implementation Description (bullets)**:
- Implement a `TaggedMetricsEventInvocationHandler` based off Tritium's `MetricsEventInvocationHandler`, though with somewhat simplified functionality
- Instrument async timelock services (`TimeLockServices`, to be clear) on the server side with two tags: client -> the relevant timelock client, and isCurrentSuspectedLeader -> boolean

**Concerns (what feedback would you like?)**:
- This encodes an assumption that there is only one leader. Though I guess we can change PaxosLeadershipCreator.isCurrentSuspectedLeader() to be a function of client name.
- I tried to preserve the existing metric names, but was there an inadvertent break here?

**Where should we start reviewing?**: TaggedMetricsEventInvocationHandler

**Priority (whenever / two weeks / yesterday)**: not so high, unless TomP thinks otherwise / we encounter serious support issues that are difficult to troubleshoot because the metrics we have aren't clear.

@tpetracca for SA in general
@schlosna for SA / sanity check on event-invocation-handler.
Thanks!

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3075)

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        markGlobalFailure();
+        if (context == null) {
+            logger.debug("Encountered null metric context likely due to exception in preInvocation: {}",
+                    UnsafeArg.of("exception", cause),
+                    cause);
+            return;
+        }
+
+        String failuresMetricName = MetricRegistry.name(getBaseMetricName(context), failuresMetricName());
+        taggedMetricRegistry.meter(MetricName.builder().safeName(failuresMetricName).build()).mark();
```
Comment (tboam) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java]: Should the failure metrics also have the tags applied?  (here, line below and in `markGlobalFailure`)

File: timelock-agent/src/main/java/com/palantir/timelock/paxos/AsyncTimeLockServicesCreator.java [hunk 2]
```diff
@@ -102,8 +103,10 @@ private static AsyncTimelockService createRawAsyncTimelockService(
...
     }
 
-    private static <T> T instrument(Class<T> serviceClass, T service, String client) {
-        // TODO(nziebart): tag with the client name, when tritium supports it
-        return AtlasDbMetrics.instrument(serviceClass, service, MetricRegistry.name(serviceClass));
+    private <T> T instrument(Class<T> serviceClass, T service, String client) {
+        return AtlasDbMetrics.instrumentWithTaggedMetrics(serviceClass, service, MetricRegistry.name(serviceClass),
+                context -> ImmutableMap.of(
+                        "client", client,
+                        "isCurrentSuspectedLeader", String.valueOf(leadershipCreator.isCurrentSuspectedLeader())));
```
Comment (tboam) [inline timelock-agent/src/main/java/com/palantir/timelock/paxos/AsyncTimeLockServicesCreator.java]: If the current node is partitioned off from the other nodes there will be a period where 2 nodes report they think they are the leader.  Don't see an easy way around this though.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-client/src/main/java/com/palantir/atlasdb/util/TaggedMetricsInvocationEventHandler.java]: Yep, good spot - that does make sense.

File: timelock-agent/src/main/java/com/palantir/timelock/paxos/AsyncTimeLockServicesCreator.java (hunk 2 above)
Comment (tboam) [reply timelock-agent/src/main/java/com/palantir/timelock/paxos/AsyncTimeLockServicesCreator.java]: And potentially something interesting to look for in the metrics.

**Comment by jeremyk-91** [comment]
I've patched an internal timelock implementation to use the new version - and observed that on leader nodes and on non-leader nodes, metrics were tagged appropriately. Furthermore, when a leader election occurred, the metric types being logged switched, which is what we want. Client names were also correctly associated with metrics.

**Pull request merged**